Feat: Custom plugin management in Konnect (#6164)
Kong/docs.konghq.com: app/konnect/gateway-manager/plugins/add-custom-plugin.md

{% endnavtab %}
{% navtab Konnect API %}

Upload the `schema.lua` file for your plugin using the `/plugin-schemas` endpoint:

```sh
curl -i -X POST \
  https://{region}.api.konghq.com/v2/{controlPlaneId}/core-entities/plugin-schemas \
  --data lua_schema=@example-schema.lua
```

This example specifies a file, but you can also include the entire schema in the request as JSON.

You should get an HTTP 201 response. You can check that your schema was uploaded using the following request:

```sh
curl -i -X GET \
  https://{region}.api.konghq.com/v2/{controlPlaneId}/core-entities/plugin-schemas
```

{% endnavtab %}
{% endnavtabs %}

Uploading a custom plugin schema adds the plugin to a specific control plane. 
If you need it to be available in multiple control planes, add the schema individually to each one.

You can now configure this custom plugin like any other plugin in {{site.konnect_short_name}}.

## Upload files to data plane nodes

After uploading a schema to {{site.konnect_short_name}}, 
upload all plugin files to **each** {{site.base_gateway}} data plane node:
* Basic modules: `schema.lua`, `handler.lua`

> * Advanced modules: `api.lua`, `daos.lua`, `migrations.lua`

Reviewer: We can remove this reference to Advanced Modules as Konnect does not support these modules (ref: https://kongstrong.slack.com/archives/C03LRB400TC/p1694035306780859)

> ---

Reviewer 2: > Someone who develops a custom plugin is already a heavy user of Gateway. The audience here is generally an experienced Kong user.

This answers my question, I think as long as the audience already knows then you don't have to add anything extra to the intro.

title: Managing custom plugins in Konnect
content_type: how-to
---

You can manage custom plugins via the {{site.konnect_short_name}} UI or 
the {{site.konnect_short_name}} Custom Plugins API.

If you need to update a schema for a plugin that has already been uploaded
to {{site.konnect_short_name}}, there are a few considerations based on the type 
of update. 
See [Editing or deleting a custom plugin's schema](/konnect/gateway-manager/plugins/update-custom-plugin/) 
for more information.

> {:.important}
> > **Caution**: Carefully test the operation of any custom plugins before deploying
> them to production. Kong is not responsible for the operation or support of any 
> custom plugins, including any performance impacts on your {{site.konnect_short_name}}
> or {{site.base_gateway}} deployments. 

Reviewer 3: ```suggestion
{:.important}
> **note**: Deploying custom plugins to production without testing them first could have unintended consequences for your system.
```

In general, this warning at the top of a doc like this makes it really uninviting for me. Like, I shouldn't do this because there is a chance I break my whole system and the org is washing their hands of any liability.

Author: I think the "washing our hands of liability" is the idea - this was a request from legal to add.

Author: I can move the warning down in the topic, but it must be here, and we can't cut it down as much as you have in the suggestion.

Using that file, it creates a plugin entry in the plugin catalog for your control plane.

Upload a custom plugin schema to create a configurable entity in {{site.konnect_short_name}}:

{% navtabs %}
{% navtab Konnect UI %}

1. From the **Gateway Manager**, open a control plane.
1. Open **Plugins** from the side navigation, then click **Add Plugin**.
1. Open the **Custom Plugins** tab, then click **Create** on the Custom Plugin tile.
1. Upload the `schema.lua` file for your plugin.
1. Check that your file displays correctly in the preview, then click **Save**.

{% endnavtab %}
{% navtab Konnect API %}

Upload the `schema.lua` file for your plugin using the `/plugin-schemas` endpoint:

```sh
curl -i -X POST \
  https://{region}.api.konghq.com/v2/{controlPlaneId}/core-entities/plugin-schemas \
  --data lua_schema=@example-schema.lua
```

This example specifies a file, but you can also include the entire schema in the request as JSON.

You should get an HTTP 201 response. You can check that your schema was uploaded using the following request:

```sh
curl -i -X GET \
  https://{region}.api.konghq.com/v2/{controlPlaneId}/core-entities/plugin-schemas
```

> 

Reviewer 3: What will I receive in that request that is different from the 201 I got in the request before? Maybe add:

"The request will return a JSON object representation of your plugin if it was updated correctly"

{% endnavtab %}
{% endnavtabs %}

Uploading a custom plugin schema adds the plugin to a specific control plane. 
If you need it to be available in multiple control planes, add the schema individually to each one.

You can now configure this custom plugin like any other plugin in {{site.konnect_short_name}}.

> ## Upload files to data plane nodes

Author: No, it is talking about data plane nodes, not configuring the plugin. To use a custom plugin, you have to upload all of its file to each node _before_ you can start configuring the plugin in Konnect.

      -e "KONG_LUA_PACKAGE_PATH=/tmp/custom_plugins/?.lua;;" \
      -e "KONG_ROLE=data_plane" \
      -e "KONG_DATABASE=off" \
      -e "KONG_VITALS=off" \
      -e "KONG_NGINX_WORKER_PROCESSES=1" \
      -e "KONG_CLUSTER_MTLS=pki" \
      -e "KONG_CLUSTER_CONTROL_PLANE=<example>.cp0.konghq.com:443" \
      -e "KONG_CLUSTER_SERVER_NAME=<example>.cp0.konghq.com" \
      -e "KONG_CLUSTER_TELEMETRY_ENDPOINT=<example>.tp0.konghq.com:443" \
      -e "KONG_CLUSTER_TELEMETRY_SERVER_NAME=<example>.tp0.konghq.com" \
      -e "KONG_CLUSTER_CERT=<cert>" \
      -e "KONG_CLUSTER_CERT_KEY=<key>" \
      -e "KONG_LUA_SSL_TRUSTED_CERTIFICATE=system" \
      -e "KONG_KONNECT_MODE=on" \
      -p 8000:8000 \
      -p 8443:8443 \
      kong/kong-gateway:<version>
    ```
1. Run the command to start a data plane node with your custom plugin loaded in.

To copy the plugin using a Dockerfile instead, see the [{{site.base_gateway}} custom plugin docs](/gateway/latest/plugin-development/distribution/).

{% endnavtab %}
{% endnavtabs %}

## More information

* [Edit or delete custom plugins in {{site.konnect_short_name}}](/konnect/gateway-manager/plugins/update-custom-plugin/)

> * [Custom plugin schema endpoints (Control Plane Config API)](/konnect/api/control-plane-configuration/latest/#/Custom%20Plugin%20Schemas)

Author: Nope - the dev portal viewer adds that second `/` if you leave it out. Technically both links will work, but the accurate one is https://docs.konghq.com/konnect/api/control-plane-configuration/latest/#/Custom%20Plugin%20Schemas.

---

> title: Managing custom plugins in Konnect

Reviewer 3: ```suggestion
title: Adding custom plugins in Konnect
```
Match file-name and nav bar

content_type: how-to
---

You can manage schemas for custom plugins via the {{site.konnect_short_name}} UI or 
the {{site.konnect_short_name}} Control Planes Config API. 

After uploading a schema to {{site.konnect_short_name}}, upload your 
custom plugin to each data plane node, then {{site.konnect_short_name}} can manage the 
configuration for your plugin like any other Kong entity.

If you need to update a schema for a plugin that was already uploaded
to {{site.konnect_short_name}}, there are a few considerations based on the type 
of update. 
See [Editing or deleting a custom plugin's schema](/konnect/gateway-manager/plugins/update-custom-plugin/) 
for more information.

## Prerequisites

> * Your custom plugin meets all [requirements](/konnect/gateway-manager/plugins/#custom-plugins)

Reviewer 3: ```suggestion
* Your custom plugin meets {{site.konnect_short_name}}'s [requirements](/konnect/gateway-manager/plugins/#custom-plugins)
```

for {{site.konnect_short_name}}.

> * The custom plugin can be written in any supported language, but the schema file must be in Lua.

Reviewer 3: ```suggestion
* The schema file must be in Lua, even if the custom plugin is written in another supported language.
```

Prereq at the front part of the sentence.

For help with developing plugins, see the [plugin development resources](#more-information).
  
    If you have a custom plugin written in a language other than Lua, convert the schema 
into a `schema.lua` file before uploading it to {{site.konnect_short_name}}.

> * If you're using the `/plugin-schemas` API, you need a personal access token or a system account
> token to authenticate with the API. You can pass your token with any API request using an 
> authentication header:

Reviewer 3: ```suggestion
* When using the `/plugin-schemas` API, authenticate your requests with either a personal access token or a system account token by including it in the authentication header:
```

Reviewer 3: just a reword feel free to ignore
